src/courses/mappers/04.md

```diff
@@ -0,0 +1,70 @@
+---
+order: 4
+next: 05.md
+title: What Is an OHDF Mapper?
+author: Charles Hu
+---
+
+## What Is a Mapper?
+
+A mapper is an example of the Adapter Design Pattern which is used to correlate (or map) items in two different objects with one another. Mappers are useful in that they allow us to correlate values in objects that are nominally different but semantically the same. A result of this is that we can easily transform one object type into another through the use of mappings which define the direct relationship of values in each object type. An important aspect of the transformation process is that while the objects containing the values are transformed, the values themselves typically remain the same.
+
+Mappings are the actual correlation of values between two different objects. They are the basis for mappers and are responsible for defining which values between two objects are semantically the same. Mappers are the implementation of mappings which perform the necessary processing to transform the objects according to relationships found in the mappings. In order to have a mapper that is as accurate as possible in its transformations, the mappings should ideally include correlations for all values in the objects.
+
+Here is a scenario which demonstrates some key aspects of mappers:
+
+::: details Transferring IDs
+Say we have a business that is changing its employee identification software and needs to transfer the current credentials of its employees from Software A to Software B. The data formats the softwares use for IDs are as follows:
+```json
+// Software A
+{Name, DoB, Age, Title}
+
+// Software B
+{employee, employeeBirthday, employeeAge, jobTitle}
+```
+
+How do we transfer John's credentials from Software A to Software B?
+```json
+{Name: 'John Doe', DoB: 10-6-1992, Age: 32, Title: 'Security Technician'}
+```
+
+What we can do is create a mapping which correlates the items from Software A's ID scheme to Software B's:
+```json
+{employee: Name, employeeBirthday: DoB, employeeAge: Age, jobTitle: Title}
+```
+
+With this, we can then develop a mapper which takes John's credentials from Software A and transforms it to Software B's format as so:
+```json
+{Name: 'John Doe', DoB: 10-6-1992, Age: 32, Title: 'Security Technician'}
+
+                                 ||
+                                \  /
+                                 \/
+
+                               MAPPER
+
+                                 ||
+                                \  /
+                                 \/
+
+{employee: 'John Doe', employeeBirthday: 10-6-1992, employeeAge: 32, jobTitle: 'Security Technician'}
+``` 
+
+The important thing to note here is that mappers rely on underlying mappings which match semantically similar fields between two objects. These matches allow us to correctly convert each item in one object to the other.
+:::
+
```

Comment:
Student interaction!  Ask a question but don't provide an answer: "Where else have you seen mappers or adapters used before (even in non-data format contexts)?"  An instructor could then discuss adapter cables ex. between VGA and HDMI.